### DISCUSSION TITLE
Material grid interpolation with conductive materals

### DISCUSSION THREAD

**Discussion by dritory** [NanoComp/meep#1997] [Q&A]
I wanted to create material grid with two different materials with different epsilons and conductivities, and I ran into some questions regarding the conductivity in Meep. 

The documentation states that interpolation is possible between two mediums with LorentzianSusceptibility, but does this imply that it will interpolate the conductivity aswell? Edit: I see that it is mentioned in the paper. Does it work in practise?

The conductivity in Meep, in relation to the SI unit conductivity, is defines as
![image](https://user-images.githubusercontent.com/183177/157864955-2d5bb2b9-a85a-4d04-957e-9c240f455a4d.png),
which means that the conductivity is relative to the real permittivity of the material. If you define two materials in Meep with the same value for the conductivity, but with different epsilon, the actual conductivity will be different by a factor of ε_r.

Now, if I want to interpolate between two materials Medium 1 and Medium 2, with different epsilon and 
the same conductivity, the actual values specified for Medium 1 and Medium 2 would differ since the conductivity is defined relative to the permittivity! How can I interpolate between two materials of different epsilon while keeping the conductivity constant? Will the conductivity remain constant if Medium 1 has a conductivity compensated by 1/ε_r1, where ε_r1 is the real permittivity of Medium 1, and likewise for Medium 2?

**Comment by smartalecH** [comment]
> How can I interpolate between two materials of different epsilon while keeping the conductivity constant?

You can't. Not unless both materials have the *same* conductivity, to begin with. There is only one degree of freedom (`u`) which is interpolating two (or more) things (permittivity, conductivity, etc.) so they have to be coupled.

The point of the implementation here is a stable interpolation method between two discrete materials for topology optimization.

For reference, here's the code that does the permittivity, conductivity, and Lorentzian interpolation:

https://github.com/NanoComp/meep/blob/ab05431dbfccc93aba57d4313ed918d9f9826b00/src/meepgeom.cpp#L561-L625

**Reply by smartalecH** [reply]
That being said, you could always *modify* the underlying interpolation scheme to fit your application (i.e. you can modify the code fragment above and build from source). Just make sure you don't have any "zero-crossings" (also discussed in our paper you referenced).

**Reply by dritory** [reply]
> You can't. Not unless both materials have the _same_ conductivity, to begin with. There is only one degree of freedom (`u`) which is interpolating two (or more) things (permittivity, conductivity, etc.) so they have to be coupled.

Does this still apply if the conductivity of the materials are equal? How can I set them to be equal if the conductivity is coupled to the permittivity?

**Reply by smartalecH** [reply]
If the conductivities are equal, interpolating between them using `u` will always yield the same result (see code fragment). Note that εᵣ and σ are each interpolated from their "extremal" values (but using the same variable `u`, of course.)

**Reply by dritory** [reply]
Agreed. The part where I'm confused is the fact that the conductivity of the medium seem to change depending on the permittivity, since σ is defined proportional to the permittivity. 

What I want is to keep one variable (σ or εᵣ) constant while interpolating the other. However, I'm concerned whether the σ actually stays the same if epsilon is interpolated. Not the Meep conductivity value, but the actual conductivity of the material. 

I ran some quick tests with the straight waveguide example with different epsilon and constant σ, and it seemed like the attenuation of the wave was different depending om the permittivity. When using a conductivity σ = σ_0 / εᵣ the attenuation was the same for different εᵣ.
